packages/block-editor/src/components/block-preview/index.js

```diff
@@ -23,22 +23,41 @@ function BlockPreview( props ) {
 	return (
 		<div className="editor-block-preview block-editor-block-preview">
 			<div className="editor-block-preview__title block-editor-block-preview__title">{ __( 'Preview' ) }</div>
-			<BlockPreviewContent { ...props } />
+			<BlockPreviewContent { ...props } srcWidth={ 280 } srcHeight={ 300 } destWidth={ 280 } />
 		</div>
 	);
 }
 
-export function BlockPreviewContent( { name, attributes, innerBlocks, settings } ) {
+export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight, destWidth } ) {
+	// Todo: It would be nice if destWidth could be calculated here, that's the only missing piece ot making this fully responsive.
+
+	// Calculate the scale factor necessary to size down the preview thumbnail.
+	const scale = Math.min( destWidth / srcWidth );
```

Comment:
Note the `destWidth` here. Right now it's being passed by the page that includes the preview component. But there's no reason this has to be. Basically if we can calculate the width necessary, we can remove that variable entirely. @youknowriad or @gziolo is this something you can help me with?

Reply:
I took a quick look at this. Presumably `destWidth` is "the current width of the preview component when rendered into the DOM`? If so then we could use `offsetWidth` to get the dimensions programmatically.

I played around to see how feasible this was. Below is the result. It's my first time using hooks so it is unlikely to be usable code so I haven't pushed it, but here's a diff of something that might help. 


```diff
diff --git a/packages/block-editor/src/components/block-preview/index.js b/packages/block-editor/src/components/block-preview/index.js
index 0af013d9d..c9608dc8a 100644
--- a/packages/block-editor/src/components/block-preview/index.js
+++ b/packages/block-editor/src/components/block-preview/index.js
@@ -1,10 +1,15 @@
 /**
  * WordPress dependencies
  */
+/**
+ * External dependencies
+ */
+import React from 'react';
 import { __ } from '@wordpress/i18n';
 import { createBlock } from '@wordpress/blocks';
 import { Disabled } from '@wordpress/components';
 import { withSelect } from '@wordpress/data';
+import { useLayoutEffect, useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -23,32 +28,48 @@ function BlockPreview( props ) {
 	return (
 		<div className="editor-block-preview block-editor-block-preview">
 			<div className="editor-block-preview__title block-editor-block-preview__title">{ __( 'Preview' ) }</div>
-			<BlockPreviewContent { ...props } srcWidth={ 280 } srcHeight={ 300 } destWidth={ 280 } />
+			<BlockPreviewContent { ...props } srcWidth={ 280 } srcHeight={ 300 } />
 		</div>
 	);
 }
 
-export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight, destWidth } ) {
+export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight } = {} ) {
 	// Todo: It would be nice if destWidth could be calculated here, that's the only missing piece ot making this fully responsive.
 
-	// Calculate the scale factor necessary to size down the preview thumbnail.
-	const scale = Math.min( destWidth / srcWidth );
-	const previewDimensions = {
-		width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
-		height: srcHeight ? srcHeight : 300 + 'px',
-		transform: 'scale(' + scale + ')',
-	};
+	const previewRef = React.createRef();
+	const [ previewDimensions, setPreviewDimensions ] = useState( {
+		width: 400,
+		height: 300,
+		transform: 'scale(1)',
+	} );
 
-	// We use a top-padding to create a responsively sized element with the same aspect ratio as the preview.
-	// The preview is then absolutely positioned on top of this, creating a visual unit.
-	const aspectPadding = Math.round( srcHeight / srcWidth * 100 );
-	const previewAspect = {
-		paddingTop: aspectPadding + '%',
-	};
+	const [ previewAspect, setPreviewAspect ] = useState( {
+		paddingTop: '75%',
+	} );
+
+	useLayoutEffect( () => {
+		const destWidth = previewRef.current.offsetWidth;
+
+		// Calculate the scale factor necessary to size down the preview thumbnail.
+		const scale = Math.min( destWidth / srcWidth ) || 1;
+
+		setPreviewDimensions( {
+			width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
+			height: srcHeight ? srcHeight : 300 + 'px',
+			transform: 'scale(' + scale + ')',
+		} );
+
+		// We use a top-padding to create a responsively sized element with the same aspect ratio as the preview.
+		// The preview is then absolutely positioned on top of this, creating a visual unit.
+		const aspectPadding = Math.round( srcHeight / srcWidth * 100 );
+		setPreviewAspect( {
+			paddingTop: aspectPadding + '%',
+		} );
+	}, [] );
 
 	const block = createBlock( name, attributes, innerBlocks );
 	return (
-		<div style={ previewAspect } className="editor-block-preview__container" aria-hidden>
+		<div ref={ previewRef } style={ previewAspect } className="editor-block-preview__container" aria-hidden>
 			<Disabled style={ previewDimensions } className="editor-block-preview__content block-editor-block-preview__content editor-styles-wrapper">
 				<BlockEditorProvider
 					value={ [ block ] }
diff --git a/packages/block-editor/src/components/block-styles/index.js b/packages/block-editor/src/components/block-styles/index.js
index a1289c27c..10ef682b3 100644
--- a/packages/block-editor/src/components/block-styles/index.js
+++ b/packages/block-editor/src/components/block-styles/index.js
@@ -133,7 +133,7 @@ function BlockStyles( {
 								innerBlocks={ block.innerBlocks }
 								srcWidth={ 400 }
 								srcHeight={ 300 }
-								destWidth={ 124 }
+
 							/>
 						</div>
 						<div className="editor-block-styles__item-label block-editor-block-styles__item-label">

```

Reply:
You know, feel free to push your changes! If it gets us closer, that's great, if not we can always revert! It's still a draft PR after all :)